### ISSUE TITLE
Feature Request JSON serializer support for AZStd::any (and AZStd::variant)

### ISSUE THREAD

**Issue by gadams3** [o3de/o3de#10828]
There are multiple, established, widely used systems like script canvas, graph canvas, graph model, material component, material editor, and previously the material system, that utilize AZStd::any to store values with multiple, dynamic data types that cannot always be predetermined or governed by the system. Many of these systems were developed before the JSON serializer was introduced and functioned as expected with the object serializer. 

Migrating to the JSON serializer requires custom serializers to be written or classes to be redesigned because there is no support for AZStd::any. Creating a custom serializer cannot be done for an individual member of a class or structure. It needs to serialize the entire class with special handling for the member using the unsupported type.

Systems like graph canvas and graph model need to be able to serialize data types specific to the tools built on top of them. The custom serializer that was introduced for graph model immediately limited its usability because it only supports data types needed by landscape canvas. That custom serialization code was copied directly from the material component. It has been copied a third time to unblock initial prototyping of material canvas dynamic node configurations.

Material Canvas, https://github.com/o3de/sig-graphics-audio/issues/51, is being built upon graph model and graph canvas that store property values using AZStd::any. Development will be blocked because it must support serializing types in the dynamic node configuration and graph model that cannot be added to all of the systems that need to contain or serialize the data.

There are multiple copies of the same JSON serializer code to support basic and math related types.

As new systems with different needs are being developed like material canvas, new types will need to be added to the custom serializers which would require those systems having knowledge of and dependencies on the libraries providing those context specific data types.

Edit and behavior context support AZStd::any and AZStd::variant

@carlitosan has already done an excellent job on the work here.
https://github.com/o3de/o3de/pull/10804

**Describe the solution you'd like**
The ideal solution would be for the JSON serializer to natively and transparently support AZStd::any. This will allow systems like script canvas, graph canvas, graph model, material canvas to build nodes that automatically support and interchange all of the necessary data types. It would also allow all of the custom serialization code to be deleted.

The same is true about AZStd::variant, used in the material system, even though it previously used AZStd::any,

**Describe alternatives you've considered**
- Creating a custom class with a custom serializer that mimics the functionality of AZStd::any then reengineering all of the systems to use it
- Converting all the data types to and from string in order to serialize them
- Reverting everything back to use the object serializer

**Additional context**
https://github.com/o3de/sig-graphics-audio/issues/51
https://github.com/o3de/o3de/pull/1224
https://github.com/o3de/o3de/blob/development/Gems/Atom/Feature/Common/Code/Source/Material/MaterialAssignmentSerializer.cpp
https://github.com/o3de/o3de/blob/development/Gems/Atom/Tools/AtomToolsFramework/Code/Source/DynamicNode/DynamicNodeSlotConfigSerializer.cpp
https://github.com/o3de/o3de/blob/development/Gems/GraphModel/Code/Source/Model/Slot.cpp#L284-L447

**Comment by lumberyard-employee-dm** [comment]
I am good with going forward with adding AZStd::any serialization support to the JSON Serialization system, based on the needs and requirements of ScriptCanvas and MaterialCanvas to be able to serialize arbitrary types
